xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java

```diff
@@ -0,0 +1,221 @@
+package info.bitrich.xchangestream.kraken;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import info.bitrich.xchangestream.core.StreamingTradeService;
+import info.bitrich.xchangestream.kraken.dto.KrakenOpenOrder;
+import info.bitrich.xchangestream.kraken.dto.KrakenOwnTrade;
+import info.bitrich.xchangestream.kraken.dto.enums.KrakenSubscriptionName;
+import info.bitrich.xchangestream.service.netty.StreamingObjectMapperHelper;
+import io.reactivex.Observable;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.Order;
+import org.knowm.xchange.dto.trade.LimitOrder;
+import org.knowm.xchange.dto.trade.MarketOrder;
+import org.knowm.xchange.dto.trade.StopOrder;
+import org.knowm.xchange.dto.trade.UserTrade;
+import org.knowm.xchange.kraken.KrakenAdapters;
+import org.knowm.xchange.kraken.dto.account.KrakenWebsocketToken;
+import org.knowm.xchange.kraken.dto.trade.KrakenOrderFlags;
+import org.knowm.xchange.kraken.dto.trade.KrakenOrderStatus;
+import org.knowm.xchange.kraken.dto.trade.KrakenType;
+import org.knowm.xchange.kraken.service.KrakenAccountServiceRaw;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+
+public class KrakenStreamingTradeService implements StreamingTradeService {
+    private static final Logger LOG = LoggerFactory.getLogger(KrakenStreamingTradeService.class);
+
+    private KrakenAccountServiceRaw rawKrakenAcctService;
+    private KrakenStreamingService streamingService;
+    private KrakenWebsocketToken token;
+    private long tokenExpires = 0L;
+
+    KrakenStreamingTradeService(KrakenStreamingService streamingService, KrakenAccountServiceRaw rawKrakenAcctService) {
+        this.streamingService = streamingService;
+        this.rawKrakenAcctService = rawKrakenAcctService;
+    }
+    private String getChannelName(KrakenSubscriptionName subscriptionName) {
+        return subscriptionName.toString();
+    }
+
+    private static class KrakenDtoOrderHolder extends HashMap<String, KrakenOpenOrder> {}
+    private static class KrakenDtoUserTradeHolder extends HashMap<String, KrakenOwnTrade> {}
+
+    private KrakenWebsocketToken renewToken() throws IOException {
+        if (System.currentTimeMillis() >= tokenExpires) {
+            token = rawKrakenAcctService.getKrakenWebsocketToken();
+            tokenExpires = System.currentTimeMillis() + (token.getExpiresInSeconds() * 900L); // 900L instead of a 1000L for 90%
+        }
+        return token;
+    }
+
+    @Override
+    public Observable<Order> getOrderChanges(CurrencyPair currencyPair, Object... args) {
+
+        return Observable.create( source -> {
```

Comment:
Why the wrapper observable? Looks like you can just return the result of `subscribeChannel`. Something like this (probably doesn't work, but certainly something like this):
```
return streamingService.subscribeChannel(getChannelName(KrakenSubscriptionName.openOrders), renewToken().getToken())
    .filter(JsonNode::isArray)
    .filter(Objects::nonNull)
    .map(jsonNode -> jsonNode.get(0))
    .map(jsonNode -> StreamingObjectMapperHelper.getObjectMapper().treeToValue(jsonNode, KrakenDtoOrderHolder[].class))
    .flatMapIterable(this::adaptKrakenOrders);
```

Reply:
that works, and I added the missing currencyPair filter

Reply:
[Kraken] review fix #3
This is not great, we get updates without CurrencyPair, so we have to let those thru,
just in case they are for our subscription  (would have to maintain order cache otherwise)

                    .filter(order -> currencyPair == null || order.getCurrencyPair() == null || order.getCurrencyPair().compareTo(currencyPair) == 0);